File: LeetCode-Java/Clone Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        HashMap<UndirectedGraphNode,UndirectedGraphNode> record = new HashMap<UndirectedGraphNode,UndirectedGraphNode>();
        return cloneGraphPart(node,record);
    }
    
    UndirectedGraphNode cloneGraphPart(UndirectedGraphNode node,HashMap<UndirectedGraphNode,UndirectedGraphNode> record)
    {
        if(node == null) return null;
        if(record.containsKey(node))
            return record.get(node);
        UndirectedGraphNode newNode = new UndirectedGraphNode(node.label);
        record.put(node, newNode);
        for(int i = 0;i < node.neighbors.size();i ++)
        {
            UndirectedGraphNode tempNode = node.neighbors.get(i);
            if(record.containsKey(tempNode))
                newNode.neighbors.add(record.get(tempNode));
            else
            {
                UndirectedGraphNode newTempNode = cloneGraphPart(tempNode,record);
                newNode.neighbors.add(newTempNode);
            }
        }
        return newNode;
    }
}

File: Leetcode-cpp/clone_graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == nullptr) return nullptr;
        step_1(node);
        step_2(node);
        int n = node->neighbors.size();
        UndirectedGraphNode* result = node->neighbors[n - 3];
        step_3(node);
        return result;
    }
    
    void step_1(UndirectedGraphNode* node) {
        if (node->neighbors.size() < 2 || node->neighbors[node->neighbors.size() - 1] != nullptr) {
            node->neighbors.push_back(new UndirectedGraphNode(node->label));
            node->neighbors.push_back(nullptr);
            for (int i = 0; i < node->neighbors.size() - 2; ++i) {
                step_1(node->neighbors[i]);
            }
        }
    }

    void step_2(UndirectedGraphNode* node) {
        if (node->neighbors.size() >= 2 && node->neighbors[node->neighbors.size() - 2] != nullptr) {
            node->neighbors.push_back(nullptr);
            UndirectedGraphNode* current = node->neighbors[node->neighbors.size() - 3];
            int n = node->neighbors.size();
            for (int i = 0; i < n - 3; ++i) {
                UndirectedGraphNode* temp = node->neighbors[i];
                int index = temp->neighbors.size() - 1;
                while (index >= 0 && temp->neighbors[index] == nullptr) index--;
                current->neighbors.push_back(temp->neighbors[index]);
            }
            for (int i = 0; i < n - 3; ++i) {
                step_2(node->neighbors[i]);
            }
        }
    }
    
    void step_3(UndirectedGraphNode* node) {
        int n = node->neighbors.size();
        if (n >= 1 && node->neighbors[n - 1] == nullptr) {
            node->neighbors.erase(node->neighbors.end() - 3, node->neighbors.end());
            for (int i = 0; i < n - 3; ++i) {
                step_3(node->neighbors[i]);
            }
        }
    }
};
